avahi-core/netlink.c

```diff
@@ -86,11 +86,12 @@ int avahi_netlink_work(AvahiNetlink *nl, int block) {
         return -1;
 
     p = (struct nlmsghdr *) nl->buffer;
+    remaining = bytes;
 
     assert(nl->callback);
 
-    for (; bytes > 0; p = NLMSG_NEXT(p, bytes)) {
-        if (!NLMSG_OK(p, (size_t) bytes)) {
+    for (; remaining > 0; p = NLMSG_NEXT(p, remaining), remaining = bytes - (((uint8_t *)p) - nl->buffer)) {
+        if (!NLMSG_OK(p, (size_t) remaining)) {
```

Comment:
Did you recognize that the NLMSG_NEXT() macro modifies its `len` argument:

```
#define NLMSG_NEXT(nlh,len)	 ((len) -= NLMSG_ALIGN((nlh)->nlmsg_len), \
...
```
So the `bytes` variable should contain the remaining size after each call of `NLMSG_NEXT()`. The [netlink(7) manpage](https://man7.org/linux/man-pages/man7/netlink.7.html) also provides an example how to use the macros:
```
           len = recvmsg(fd, &msg, 0);

           for (nh = (struct nlmsghdr *) buf; NLMSG_OK (nh, len);
                nh = NLMSG_NEXT (nh, len)) {
               /* The end of multipart message */
               if (nh->nlmsg_type == NLMSG_DONE)
                   return;

               if (nh->nlmsg_type == NLMSG_ERROR)
                   /* Do some error handling */
               ...

               /* Continue with parsing payload */
               ...
           }
```

So is there a benefit using (hard to understand) pointer arithmetic instead?